File: server/src/main/kotlin/plugins/Log.kt
/*
 * 🐻‍❄️📦 charted-server: Free, open source, and reliable Helm Chart registry made in Kotlin.
 * Copyright 2022-2023 Noelware, LLC. <team@noelware.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.noelware.charted.server.plugins

import dev.floofy.utils.koin.inject
import dev.floofy.utils.kotlin.doFormatTime
import dev.floofy.utils.slf4j.logging
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.application.hooks.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.util.*
import io.prometheus.client.Histogram
import io.sentry.Sentry
import org.apache.commons.lang3.time.StopWatch
import org.noelware.charted.Server
import org.noelware.charted.common.extensions.closeable.closeQuietly
import org.noelware.charted.common.extensions.formatting.doFormatTime
import org.noelware.charted.modules.metrics.MetricsSupport
import org.noelware.charted.modules.metrics.prometheus.PrometheusMetricsSupport
import org.noelware.charted.modules.tracing.Tracer
import org.noelware.charted.modules.tracing.Transaction
import org.noelware.charted.server.internal.DefaultServer
import org.noelware.charted.server.internal.bootTime
import org.noelware.charted.server.internal.hasStarted
import org.slf4j.MDC
import java.lang.IllegalStateException

object Log: BaseApplicationPlugin<ApplicationCallPipeline, Unit, Log> {
    private val histogramKey: AttributeKey<Histogram.Timer> = AttributeKey("Prometheus Histogram Timer")
    private val transactionKey: AttributeKey<Transaction> = AttributeKey("Transaction")
    private val stopwatchKey: AttributeKey<StopWatch> = AttributeKey("Stopwatch")
    private val log by logging("org.noelware.charted.server.plugins.KtorLogger")

    private val metrics: MetricsSupport by inject()
    private val tracer: Tracer? = Tracer.globalOrNull()
    private val server: Server by inject()

    override val key: AttributeKey<Log> = AttributeKey("Logging Plugin")
    override fun install(pipeline: ApplicationCallPipeline, configure: Unit.() -> Unit): Log {
        install(pipeline)
        return this
    }

    private fun install(pipeline: ApplicationCallPipeline) {
        val environment = pipeline.environment!!
        environment.monitor.subscribe(ApplicationStarted, this::onApplicationStarted)
        environment.monitor.subscribe(ApplicationStopped, this::onApplicationStopped)

        // equivalent: onCall { call -> }
        pipeline.intercept(ApplicationCallPipeline.Plugins) {
            call.attributes.put(stopwatchKey, StopWatch.createStarted())
            (server as DefaultServer).requestsHandled.incrementAndGet()

            MDC.put("http.method", call.request.httpMethod.value)
            MDC.put("http.version", call.request.httpVersion)
            MDC.put("http.url", call.request.path())

            val transaction = tracer?.createTransaction("${call.request.httpVersion} ${call.request.httpMethod.value.uppercase()} ${call.request.path()}", "HTTP Request")
            if (transaction != null) call.attributes.put(transactionKey, transaction)
            if (metrics is PrometheusMetricsSupport) {
                val m = metrics as PrometheusMetricsSupport

                m.serverRequests.inc()
                call.attributes.put(
                    histogramKey,
                    m.serverRequestLatency.labels(
                        call.request.httpMethod.value,
                        call.request.path(),
                        call.request.httpVersion,
                    ).startTimer(),
                )
            }
        }

        // equivalent: on(ResponseSent) { call -> }
        pipeline.sendPipeline.intercept(ApplicationSendPipeline.Engine) {
            val method = call.request.httpMethod
            val version = call.request.httpVersion
            val endpoint = call.request.path()
            val status = call.response.status() ?: HttpStatusCode(-1, "Unknown HTTP Method")
            val histogram = call.attributes.getOrNull(histogramKey)
            val stopwatch = call.attributes[stopwatchKey]
            val transaction = call.attributes.getOrNull(transactionKey)
            val userAgent = call.request.userAgent() ?: "Unknown"

            // Fixes with issues that might occur when handling a request.
            // I blame Ktor for being weird :(
            val formattedTime = run {
                try {
                    stopwatch.stop()
                    " (${stopwatch.doFormatTime()})"
                } catch (e: IllegalStateException) {
                    if (e.message?.contains("Stopwatch is not running.") == false) {
                        throw e
                    }

                    ""
                }
            }

            transaction?.closeQuietly()
            histogram?.observeDuration()
            log.info("~> ${method.value} $endpoint <$version> ~ ${status.value} ${status.description} [$userAgent${formattedTime.ifBlank { "" }}]".trim())

            MDC.remove("http.method")
            MDC.remove("http.version")
            MDC.remove("http.url")

            if (Sentry.isEnabled()) Sentry.setUser(null)
        }
    }

    private fun onApplicationStarted(@Suppress("UNUSED_PARAMETER") app: Application) {
        val time = (System.nanoTime() - bootTime).doFormatTime()

        log.info("API server has started in $time")
        hasStarted.value = true
    }

    private fun onApplicationStopped(app: Application) {
        app.environment.monitor.unsubscribe(ApplicationStarted, this::onApplicationStarted)
        app.environment.monitor.unsubscribe(ApplicationStopped, this::onApplicationStopped)

        log.warn("API server has successfully shut down!")
        hasStarted.value = false
    }
}

File: server/src/middleware/log.rs
// 🐻‍❄️📦 charted-server: Free, open source, and reliable Helm Chart registry made in Rust
// Copyright 2022-2023 Noelware, LLC. <team@noelware.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::{metrics::REQUEST_LATENCY_HISTOGRAM, Server};
use axum::{
    extract::{FromRequestParts, State},
    http::{header::USER_AGENT, HeaderMap, Method, Request, Uri, Version},
    middleware::Next,
    response::IntoResponse,
};
use std::{sync::atomic::Ordering, time::Instant};

#[derive(FromRequestParts)]
pub struct Metadata {
    pub(crate) uri: Uri,
    pub(crate) method: Method,
    pub(crate) version: Version,
    pub(crate) headers: HeaderMap,
}

pub async fn log<B>(
    metadata: Metadata,
    State(server): State<Server>,
    req: Request<B>,
    next: Next<B>,
) -> impl IntoResponse {
    let histogram = &*REQUEST_LATENCY_HISTOGRAM;
    server.requests.fetch_add(1, Ordering::SeqCst);

    let uri = metadata.uri.path();
    if uri.contains("/heartbeat") {
        return next.run(req).await;
    }

    let start = Instant::now();
    let method = metadata.method.as_str();
    let version = match metadata.version {
        Version::HTTP_09 => "http/0.9",
        Version::HTTP_10 => "http/1.0",
        Version::HTTP_11 => "http/1.1",
        Version::HTTP_2 => "http/2.0",
        Version::HTTP_3 => "http/3.0",
        _ => "http/???",
    };

    let ua = metadata
        .headers
        .get(USER_AGENT)
        .map(|f| String::from_utf8_lossy(f.as_bytes()).to_string());

    let http_span = info_span!(
        "http.request",
        req.ua = ua,
        http.uri = uri,
        http.method = method,
        http.version = version
    );

    let _guard = http_span.enter();
    info!(
        http.uri = uri,
        http.method = method,
        http.version = version,
        req.ua = ua,
        "processing request"
    );

    let res = next.run(req).await;
    let now = start.elapsed();
    let status = res.status().as_u16();
    histogram.observe(now.as_secs_f64());

    info!(
        http.uri = uri,
        http.method = method,
        http.version = version,
        http.status = status,
        http.latency = format!("{now:?}").as_str(),
        req.ua = ua,
        "processed request"
    );

    res
}
